### ISSUE TITLE
HDFormer: High-order Directed Transformer for 3D Human Pose Estimation

### ISSUE THREAD

**Issue by gwleeee** [gwleeee/PaperReview#8]
# HDFormer: High-order Directed Transformer for 3D Human Pose Estimation

https://arxiv.org/abs/2302.01825
https://github.com/hyer/HDFormer

## Abstract
- 기존 pose estimation 방법들에서는 human body joints들의 관계만을 학습
- joint overlapping이 발생하거나, 포즈가 빠르게 변화할때 추론 실패율이 상승함
- High-order Directed Transformer (HDFormer) 제안
- bone과 관절의 관련성을 일관되게 활용하여 성능을 향상
- Transformer에서는 self-attention / high-order attention 방법을 채택 → multi-order attention
![image](https://github.com/gwleeee/PaperReview/assets/83268025/efdff029-3f7e-4d81-8e53-7e7fa2d54ead)

  - first-order : joint ↔ joint
  - second-order : bone ↔ joint
  - high-order : hyperbone(jointset) ↔ joint
- hard case (포즈가 빠르게 변하는 경우, 오클루전 심한 경우)에서 모션 보상을 잘 해줌
- 기존 sota 모델들보다 1/10 크기의 파라미터를 가짐

## Proposed method
![image](https://github.com/gwleeee/PaperReview/assets/83268025/91b21b6e-a7fd-4618-a3b7-f4ead9ea87b3)

### High-order Directed Transformer

- **First-order attention**
![image](https://github.com/gwleeee/PaperReview/assets/83268025/5113c120-8961-48d1-91a1-8aeb293c8329)    
    1. 여기서 linear는 그냥 linear가 아니고, 공간 채널을 1x1 커널의 컨볼루션으로 섞은 다음, 시간축으로 펼침
        - (N C T V) → conv → (N C’ T V) → rearrange → (N **V C’*T**)
        - (N C T V) → conv → (N C’ T V) → (N **C’*T V**)
    2. 위 두 텐서를 곱하여 (N V V) 텐서 획득 후 softmax로 joint를 확률변수화 함
    3. 확률변수화된 joint 정보에 대하여 인접행렬과 러너블 파라미터 phi의 합을 더함
        1. 즉 3번까지의 결과는 인접한 joint 정보와, 학습된 joint 관계, joint의 attention 정보를 합한 (V V) 어텐션 매트릭스임
    4. joint feature와 어텐션 매트릭스를 곱하고 residual을 더하여 유효한 joint 성분을 강화
    - **First order attention에서는 인접한/학습을 통해 관계된 joint 성분들을 결합하여 attention하는 역할을 함**

- **Hyperbone Representation**
![image](https://github.com/gwleeee/PaperReview/assets/83268025/164556e4-f80b-4647-a27c-d46d5837aeb4)    
    - Graph내 특정 joint 쌍에 대하여 최단거리일때 거쳐가는 joint의 집합들 → hyperbone
        - (0,3)→ {0,1,2,3}
        - (0, 16) → {0,7,8,14,15,16}
    - hyperbone들은 joint feature selection 과정을 거쳐서 corresponding feature H로 표현됨
        - H_{0,3} = {z_0, z_1, z_2, z_3}
            - z는 joint feature
    - instantiation function
        - 이 함수는 여러 기능으로 정의 가능함
            1. 빼기연산 : hyperbone의 시점과 종점의 feature를 뺀 후 linear 연산
            2. 합/곱 : hyperbone의 joint feature들을 평균내거나 곱한 후 linear 연산
            3. concatenation : hyperbone의 joint feature들을 나열한 후 linear 연산
            4. sub-concat : hyperbone의 연속적인 두 joint feature를 빼서 나열한 후 linear 연산
    - **즉, hyperbone 표현에서는, 특정 joint 사이의 최단거리로 거쳐가는 joint들의 feature들을 여러 방법으로 결합하여 joint 사이의 관계로 표현해 주는 역할을 함**

- **High-order Directed Transformer**
![image](https://github.com/gwleeee/PaperReview/assets/83268025/4a0fa3e9-c00c-43c8-96c9-29acb481612a)    
![image](https://github.com/gwleeee/PaperReview/assets/83268025/cada56cd-acb3-4b80-88ee-ff8f6d07ff3f)    
    - Query → Attention 하기 위한 대상
        - first-order attention 결과
        - Transformer의 attention을 계산하기 위한 요소로 사용
    - Key → Attention에 필요한 Query의 비교대상
        - Hyperbone 표현에 weight를 곱한 결과
        - Query와의 내적을 통해 transformer의 attention을 계산하기 위해 사용
        - 내적에 사용되므로 query와 유사도를 측정하는 값
    - Value → Attention에 따라 추출할 정보
        - Hyperbone 표현에 weight를 곱한 결과
    - 작동 순서
        1. Query와 Key를 내적한 후 softmax를 통해 정규화하여 Attention 분포 계산
        2. Value와 attention을 곱하여 최종 값을 계산 (정보 추출)
    - 정리하자면,
        - **First-order attention을 통해 얻은 시퀀스 정보에서, hyperbone과 관련성 있는 정보를 추출하는 과정**
- Loss
    - MPJPE, Motion loss
    
### Network architecture
![image](https://github.com/gwleeee/PaperReview/assets/83268025/30b5f9b5-717b-4aae-8cd5-0eee9dccae5c)
- Downsampling - upsampling 과정에서는 First-order Attention만 진행
- 이후 merging stage에서 HDFormer 연산

## Experiment
- AdaMod optimizer
- lr : 0.005 / 특정 epoch에서 1/10로 감소
- weight decay : 0.00005
- dropout : 0.3
- 슬라이딩 윈도우 기법을 이용하여 입력 시퀀스 길이를 일치시킴
![image](https://github.com/gwleeee/PaperReview/assets/83268025/633d46f3-9d01-415a-9f27-b418d13dbde8)
![image](https://github.com/gwleeee/PaperReview/assets/83268025/d22df221-a0cf-42aa-a4df-777c8486d167)
